File: Server/src/main/content/global/skill/production/herblore/potions/GenericPotion.java
package content.global.skill.production.herblore.potions;

import core.game.node.item.Item;

/**
 * The Generic potion.
 */
public final class GenericPotion {

    private final Item base;

    private final Item ingredient;

    private final int level;

    private final double experience;

    private final Item product;

    /**
     * Instantiates a new Generic potion.
     *
     * @param base       the base
     * @param ingredient the ingredient
     * @param level      the level
     * @param experience the experience
     * @param product    the product
     */
    public GenericPotion(final Item base, final Item ingredient, final int level, final double experience, final Item product) {
        this.base = base;
        this.ingredient = ingredient;
        this.level = level;
        this.experience = experience;
        this.product = product;
    }

    /**
     * Transform generic potion.
     *
     * @param potion the potion
     * @return the generic potion
     */
    public static GenericPotion transform(final UnfinishedPotion potion) {
        return new GenericPotion(potion.getBase(), potion.getIngredient(), potion.getLevel(), 0, potion.getPotion());
    }

    /**
     * Transform generic potion.
     *
     * @param potion the potion
     * @return the generic potion
     */
    public static GenericPotion transform(final FinishedPotion potion) {
        return new GenericPotion(potion.getUnfinished().getPotion(), potion.getIngredient(), potion.getLevel(), potion.getExperience(), potion.getPotion());
    }

    /**
     * Gets base.
     *
     * @return the base
     */
    public Item getBase() {
        return base;
    }

    /**
     * Gets level.
     *
     * @return the level
     */
    public int getLevel() {
        return level;
    }

    /**
     * Gets experience.
     *
     * @return the experience
     */
    public double getExperience() {
        return experience;
    }

    /**
     * Gets product.
     *
     * @return the product
     */
    public Item getProduct() {
        return product;
    }

    /**
     * Gets ingredient.
     *
     * @return the ingredient
     */
    public Item getIngredient() {
        return ingredient;
    }
}

File: Server/src/main/content/global/skill/production/herblore/data/potion/GenericPotion.kt
package content.global.skill.production.herblore.data.potion

import core.game.node.item.Item

/**
 * Generic potion
 *
 * @property base
 * @property ingredient
 * @property level
 * @property experience
 * @property product
 * @constructor Generic potion
 */
class GenericPotion(
    val base: Item?,
    val ingredient: Item?,
    val level: Int,
    val experience: Double,
    val product: Item?
) {
    companion object {
        fun transform(potion: UnfinishedPotion): GenericPotion {
            return GenericPotion(
                base = potion.base,
                ingredient = potion.ingredient,
                level = potion.level,
                experience = 0.0,
                product = potion.potion
            )
        }

        fun transform(potion: FinishedPotion): GenericPotion {
            return GenericPotion(
                base = potion.unfinished.potion,
                ingredient = potion.ingredient,
                level = potion.level,
                experience = potion.experience,
                product = potion.potion
            )
        }
    }
}
